lvsfunc/comparison.py

```diff
@@ -27,111 +27,64 @@ class Direction(IntEnum):
-                             "consider resampling to YUV444 or RGB before attempting to crop")
-
-        if direction == Direction.HORIZONTAL:
-            crop_width, overflow = self.width // self.num_clips, self.width % self.num_clips
-
-            for key, (name, clip) in enumerate(self.clips.items()):
-                left_crop = crop_width * key
-                right_crop = crop_width * (self.num_clips - 1 - key)
-
-                if key != (self.num_clips - 1):
-                    right_crop += overflow
+        self.clips: List[vs.VideoNode] = list(clips.values()) if isinstance(clips, dict) else list(clips)
+        self.names: Optional[List[str]] = list(clips.keys()) if isinstance(clips, dict) else None
 
-                self.clips[name] = clip.std.Crop(left=left_crop, right=right_crop)
+        self.label_alignment = label_alignment
 
-        elif direction == Direction.VERTICAL:
-            crop_height, overflow = self.height // self.num_clips, self.height % self.num_clips
+        self.num_clips = len(clips)
 
-            for key, (name, clip) in enumerate(self.clips.items()):
-                top_crop = crop_height * key
-                bottom_crop = crop_height * (self.num_clips - 1 - key)
+        widths: Set[int] = {clip.width for clip in self.clips}
+        self.width: Optional[int] = w if (w := widths.pop()) != 0 and len(widths) == 0 else None
 
-                if key != (self.num_clips - 1):
-                    bottom_crop += overflow
+        heights: Set[int] = {clip.height for clip in self.clips}
+        self.height: Optional[int] = h if (h := heights.pop()) != 0 and len(heights) == 0 else None
 
-                self.clips[name] = clip.std.Crop(top=top_crop, bottom=bottom_crop)
+        formats: Set[Optional[str]] = {getattr(clip.format, 'name', None) for clip in self.clips}
+        self.format: Optional[str] = formats.pop() if len(formats) == 0 else None
 
-    def _markclips(self) -> List[vs.VideoNode]:
+    def _markclips(self) -> None:
         """
-        This allows for internal functions to specify a space-only string,
-        such as '   ' that avoid placing a label on the clip.
-
-        By default, `vapoursynth.core.text.Text(text='')` will not place a label,
-        but specifying multiple clips in the **clips Dict with the same '' key will not work.
+        If a `name` is only space characters, `'   '`, for example, the name will not be overlayed on the clip.
         """
-        clips: List[vs.VideoNode] = []
-        for name, clip in self.clips.items():
-            if name.strip():
-                clips.append(clip.text.Text(text=name, alignment=self.label_alignment))
-            else:
-                clips.append(clip)
-
-        return clips
+        if self.names:
+            self.clips = [clip.text.Text(text=name, alignment=self.label_alignment) if name.strip() else clip for clip, name in zip(self.clips, self.names)]
```

Comment:
Why did you change this method to mutate `self.clips`?

Reply:
I think originally I was using this to return a list of clips as I was essentially `list()`-ing an items view. There's no point in it returning anything now since clips is always a list.

Reply:
From experience, dealing with state on an instance when the state is only used at one point in the pipeline (in `.compare`), you're 

1. pretending that there is internal state that persists across runs, which it does not,
2. causing inconsistencies of state to become more likely.

Unless I misremember things from the aruably really quick review from last night.

Reply:
So I should change it back to how it was originally? I only understand a portion of your comment but don't see how it applies here.

Reply:
Yes. I'll try to explain it differently:

The implemented classes (following the `Comparer` interface) currently do the following:
1. take a few parameters as input
2. produce a clip as a direct result of the input

In a sense, they are functions with inputs and a single output. This is a linear process.

They are implemented as classes to make use of polymorphism, which is a reasonable use case. Since they are a class, the inputs and some intermediate states used to calculate the output can also be accessed by callers (i.e. `self.width`), but that's more of an implementation through polymorphism. However, `self.clips` is mutated, which means even though it *could* be used to refer to the input variables, this is only true until the first access of `clip` (which is a property and does not indicate mutation of the class's state), after which `self.clips` refers to whatever `_markclips` did with them. Additionally, this intermediate state isn't cached or checked within the implementation itself, so accessing `clip` twice will reapply `_markclips`.

There are a few solutions to this:

1. Make `_markclips` return the modified clips instead of writing to `self.clips`.
2. Cache the result of `clip` with `functools.cached_property`. This should be documented.
3. Change `clip` into a classmethod that forwards its args to the constructor and then returns the result of its `_compare` (which might as well be inlined).
4. Implement the comparers as classmethods of `Comparer` and have them used as `Comparer.stack(...)` or `Comparer.interleave(...)`, returning the resulting clip.

All of these have their advantages and some only address part of the issues I mentioned, while others cover them up. In the end, I think what influences the decision the most is the answer to the question "What should instances of `Comparer` be used for?".

Reply:
Addressed. Split's `_smart_crop()` *has* to mutate `self.clips` though, I tried changing it and it would require re-writing it so much that there would be no point of it inheriting `Stack`, nor `Comparer`.